### PR TITLE
Merge gsplat (without adaptive load balancing)

### PR DESCRIPTION
**Pull request by alexis-mmm**
1. Support batched kernels in the preprocess stage.
2. Use bsz=1 in the render stage due to various numbers of gaussians.

### CODE REVIEW

File: gaussian_renderer/__init__.py [hunk 1]
```diff
@@ -766,9 +1027,168 @@ def render_final(batched_screenspace_pkg, batched_strategies):
...
 
+    ########## [END] CUDA Rasterization Call ##########
+    return batched_rendered_image, batched_compute_locally
+  
+def gsplat_render_final(batched_screenspace_pkg, batched_strategies, tile_size=16):
+    """
+    Render the scene. 
+    """
+    timers = utils.get_timers()
+    C = len(batched_strategies)
```
Comment (TarzanZhao) [inline gaussian_renderer/__init__.py:1043]: C is batch size(num of cameras), right? When I first read this, I thought this is batch size which means we should use B (rather than C). However, gsplat uses C for batch size(num of cameras) here. What do you think is better to use here?

**Review by TarzanZhao — COMMENTED** [review]
I think this is a great work! I have reviewed the distribution logic of this code and I have left a few comments. I did not see big problems in your code. And I will review few remaining details related to gsplat api usage on Wednesday.

**Comment by TarzanZhao** [comment]
I have finished the review. I do not have other comments. We can merge this into main branch after solving this conflicting file: .gitmodules

File: gaussian_renderer/__init__.py [hunk 2]
```diff
@@ -712,7 +815,165 @@ def distributed_preprocess3dgs_and_all2all_final(batched_viewpoint_cameras, pc :
...
+        focal_length_y = viewpoint_camera.image_height / (2 * tanfovy)
+        K = torch.tensor(
+            [
+                [focal_length_x, 0, viewpoint_camera.image_width / 2.0],
+                [0, focal_length_y, viewpoint_camera.image_height / 2.0],
+                [0, 0, 1],
+            ],
+            device="cuda"
+        )
+        viewmat = viewpoint_camera.world_view_transform.transpose(0, 1) # why transpose
```
[the remark this answers is not in the document]
Comment (alexis-mmm) [reply gaussian_renderer/__init__.py:873]: Got that!

File: gaussian_renderer/__init__.py (hunk 1 above)
Comment (alexis-mmm) [reply gaussian_renderer/__init__.py:1043]: I guess C is for "cameras" in gsplat and I just follow their style so it's easier to code. But I agree that using B here causes fewer misunderstandings. Will fix it.

File: scene/gaussian_model.py [hunk 3]
```diff
@@ -716,6 +716,14 @@ def densify_and_prune(self, max_grad, min_opacity, extent, max_screen_size):
     def add_densification_stats(self, viewspace_point_tensor, update_filter):# the :2] is a weird implementation. It is because viewspace_point_tensor is (N, 3) tensor.
         self.xyz_gradient_accum[update_filter] += torch.norm(viewspace_point_tensor.grad[update_filter,:2], dim=-1, keepdim=True)
         self.denom[update_filter] += 1
+    
+    def gsplat_add_densification_stats(self, viewspace_point_tensor_grad, update_filter, width, height):# the :2] is a weird implementation. It is because viewspace_point_tensor is (N, 3) tensor.
+        grad = viewspace_point_tensor_grad # (N, 2)
+        # Normalize the gradients to [-1, 1] screen size
```
[the remark this answers is not in the document]
Comment (alexis-mmm) [reply scene/gaussian_model.py:722]: I kept this comment when reading gsplat's codes and I have the same confusion as you. In gsplat [simple trainer](https://github.com/nerfstudio-project/gsplat/blob/e62331179b6b42e7a2df4dc6b675174f186c030f/examples/simple_trainer.py#L674) they do this scaling too in their own pipeline with similar comments.  I looked into the `means2D` data (not grad) and they're not normalized, exactly the same as if using `diff-gaussian-rasterization`. It seems that the evaluating psnr will drop without  scaling the grad, so it's necessary for a correct implementation.

[the remark this answers is not in the document]
Comment (TarzanZhao) [reply scene/gaussian_model.py:722]: Get it! Let me look into it later

**Comment by alexis-mmm** [comment]
All fixed. The conflict is due to url changing of the Grendel repo and should not be a big deal.
